presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/JoinSwappingUtils.java

```diff
@@ -0,0 +1,173 @@
+import com.facebook.presto.sql.planner.optimizations.StreamPropertyDerivations;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.getTaskConcurrency;
+import static com.facebook.presto.SystemSessionProperties.isJoinSpillingEnabled;
+import static com.facebook.presto.SystemSessionProperties.isSpillEnabled;
+import static com.facebook.presto.sql.planner.optimizations.StreamPreferredProperties.defaultParallelism;
+import static com.facebook.presto.sql.planner.optimizations.StreamPreferredProperties.exactlyPartitionedOn;
+import static com.facebook.presto.sql.planner.optimizations.StreamPreferredProperties.fixedParallelism;
+import static com.facebook.presto.sql.planner.optimizations.StreamPreferredProperties.singleStream;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.gatheringExchange;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.systemPartitionedExchange;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class JoinSwappingUtils
+{
+    private JoinSwappingUtils() {}
+
+    public static Optional<JoinNode> createRuntimeSwappedJoinNode(
+            JoinNode joinNode,
+            Metadata metadata,
+            SqlParser parser,
+            Lookup lookup,
+            Session session,
+            VariableAllocator variableAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        JoinNode swapped = joinNode.flipChildren();
+
+        PlanNode newLeft = swapped.getLeft();
+        Optional<VariableReferenceExpression> leftHashVariable = swapped.getLeftHashVariable();
+        // Remove unnecessary LocalExchange in the current probe side. If the immediate left child (new probe side) of the join node
+        // is a localExchange, there are two cases: an Exchange introduced by the current probe side (previous build side); or it is a UnionNode.
+        // If the exchangeNode has more than 1 sources, it corresponds to the second case, otherwise it corresponds to the first case and could be safe to remove
+        PlanNode resolvedSwappedLeft = lookup.resolve(newLeft);
+        if (resolvedSwappedLeft instanceof ExchangeNode && resolvedSwappedLeft.getSources().size() == 1) {
+            // Ensure the new probe after skipping the local exchange will satisfy the required probe side property
+            if (checkProbeSidePropertySatisfied(resolvedSwappedLeft.getSources().get(0), metadata, parser, lookup, session, variableAllocator)) {
+                newLeft = resolvedSwappedLeft.getSources().get(0);
+                // The HashGenerationOptimizer will generate hashVariables and append to the output layout of the nodes following the same order. Therefore,
```

Comment:
Can you add a test with and without optimize_hash_generation? We can also disable hashgenerationoptimizer for presto-on-spark - i'm not sure how much value it will provide

Reply:
added to the end-to-end tests in TestPrestoSparkAdaptiveJoinQueries